pkg/util/certificates/csr.go

```diff
@@ -38,3 +45,86 @@ func ParseCertificateRequestObject(obj *certificates.CertificateSigningRequest)
 	}
 	return csr, nil
 }
+
+// NewCertificateRequest generates a PEM-encoded CSR using the supplied private
+// key data, subject, and SANs. If the private key data is empty, it generates a
+// new ECDSA P256 key to use and returns it together with the CSR data.
+func NewCertificateRequest(keyData []byte, subject *pkix.Name, dnsSANs []string, ipSANs []net.IP) (csr []byte, key []byte, err error) {
+	var privateKey interface{}
+	var privateKeyPemBlock *pem.Block
+
+	if len(keyData) == 0 {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		ecdsaKey := privateKey.(*ecdsa.PrivateKey)
+		derBytes, err := x509.MarshalECPrivateKey(ecdsaKey)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		privateKeyPemBlock = &pem.Block{
+			Type:  "EC PRIVATE KEY",
+			Bytes: derBytes,
+		}
+	} else {
+		privateKeyPemBlock, _ = pem.Decode(keyData)
```

Comment:
decoding should loop on PEM blocks found in the keyData until it finds a block it recognizes or exhausts the file. see https://github.com/liggitt/kubernetes/blob/tls-bootstrap/pkg/util/certificates/csr.go#L68-L86